src/android/SecureStorage.java

```diff
@@ -66,192 +140,340 @@ public boolean execute(String action, CordovaArgs args, final CallbackContext ca
+        editor.putString(MIGRATED_FOR_SECURITY, "TRUE");
+        editor.commit();
+    }
 
-        // Check if there is a lock screen that requires authentication defined
-        // It gives the user the possibility of defining one if there isn't one
-        // Used by the Ciphered Local Storage Plugin at startup
-        if ("secureDevice".equals(action)) {
-            Log.v(TAG, "Called secureDevice action");
-            handleLockScreen(IntentRequestType.SECURE_DEVICE, null, callbackContext);
-            return true;
+    private int initializePreferences() {
+
+        Context ctx = getContext();
+        File prefdir = new File(ctx.getApplicationInfo().dataDir,"shared_prefs");
+        String[] filenames = prefdir.list();
+        int i = 0;
+        for(String name : filenames){
+            if(name.contains("SS")){
+
+                String alias = name.substring(0, name.length() - 4);
+                String service = alias.substring(ctx.getPackageName().length() + 1, alias.length() -3);
+
+                SharedPreferencesHandler PREFS = new SharedPreferencesHandler(alias, getContext());
+                putStorage(service, PREFS);
+                i+=1;
+            }
         }
+        return i;
+    }
+
+    // Store a key/enc-value pair in SharedPreferences
+    // The encryption uses the key with an alias associated with the store name
+    private boolean set(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
+        Log.v(TAG, "Called set action");
+        final String service = args.getString(0);
+        final String key = args.getString(1);
+        final String value = args.getString(2);
 
-        // The remaining actions are the SharedPreferences interface
-        if ("remove".equals(action)) {
-            Log.v(TAG, "Called remove action");
-            String service = args.getString(0);
-            String key = args.getString(1);
-            getStorage(service).remove(key);
+        String result = encrytionHelper(service, key, value);
+
+        if(!result.contains("ERROR")){
             callbackContext.success();
-            return true;
         }
+        else{
+            callbackContext.error(result);
+        }
+        return true;
 
-        if ("store".equals(action)) {
-            Log.v(TAG, "Called store action");
-            String service = args.getString(0);
-            String key = args.getString(1);
-            String value = args.getString(2);
-            getStorage(service).store(key, value);
-            callbackContext.success();
-            return true;
+    }
+
+    private String encrytionHelper(String service, String key, String value) {
+
+        String result = "";
+
+        EncryptionExecutor encryptionExecutor = new EncryptionExecutor(service, key, value, cordova.getContext());
+        Future<String> exec = cordova.getThreadPool().submit(encryptionExecutor);
+
+        try{
+            result = exec.get();
```

Comment:
I'm not used to using Future but something feels off here. We're deferring the execution of the encryption to another thread, but then we're waiting for the result in the original thread. This is blocking the original thread, because it will be stopped until the result is ready.

Is this what's intended? And if yes, what do we gain from it, when compared to calling directly in the original thread?

Reply:
The previous version also launched in the main thread pool of cordova the encryption/decryption functions.
In this case in order for the migration to work as smooth as possible the results have to be waited in orded not to save wrong values or lose stored information in the process